### PR TITLE
Updating Omniledger to behave

### PR DESCRIPTION
**Pull request by ineiti**
There were a couple of things astray in Omniledger land:
- verification of signatures on instructions were done _after_ the contract has been run
  -> verification of signatures is done in contracts, so they can do whatever seems fit
- instructions sent to an instance didn't call that instance if it was a 'spawn' instruction
  -> all instructions call now the instance in the 'InstanceID' field
  -> fixed tests that sent instructions to non-existent instances
- followers didn't verify the signatures of instructions
  -> verification is done in contracts, so this is taken care of
- updated java and added `deriveID` again

### CODE REVIEW

File: omniledger/service/service.go [hunk 1]
```diff
@@ -768,10 +675,9 @@ func (s *Service) createStateChanges(coll *collection.Collection, scID skipchain
 	// ignore the error and compute the state changes.
 	merkleRoot, ctsOK, ctsBad, states, err = s.stateChangeCache.get(scID, cts.Hash())
 	if err == nil {
-		log.Lvl4(s.ServerIdentity(), "state changes from cache: HIT")
+		log.Lvl3(s.ServerIdentity(), "loaded state changes from cache")
 		return
 	}
-	log.Lvl4(s.ServerIdentity(), "state changes from cache: MISS")
```
Comment (kc1212) [inline omniledger/service/service.go]: I think it's useful to keep these when debugging cache problems.

File: omniledger/service/service_test.go [hunk 2]
```diff
@@ -464,57 +464,64 @@ func TestService_StateChange(t *testing.T) {
 			return nil, nil, err
 		}
 
+		zeroBuf := make([]byte, 8)
+		switch {
 		// create the object if it doesn't exist
-		if !rec.Match() {
-			if inst.Spawn == nil {
-				return nil, nil, errors.New("expected spawn")
+		case inst.Spawn != nil:
```
Comment (jeffallen) [inline omniledger/service/service_test.go]: Why not convert this one to switch `inst.GetType()` as well?

File: omniledger/service/transaction.go [hunk 3]
```diff
@@ -143,6 +136,11 @@ func (instr Instruction) GetContractState(coll CollectionView) (contractID strin
 	var cv []interface{}
 	cv, err = record.Values()
 	if err != nil {
+		if bytes.Compare(ConfigInstanceID[:], instr.InstanceID[:]) == 0 {
```
[the remark this answers is not in the document]
Comment (ineiti) [reply omniledger/service/transaction.go]: I actually thought: "why doesn't jeff like `.Equal`?" when writing this. +1

**Comment by jeffallen** [comment]
This might be failing for the same reason as #1421 , and the latest fix I made there worked. So once it is checked in, this might get fixed.

File: external/java/src/main/java/ch/epfl/dedis/lib/omniledger/Instruction.java [hunk 4]
```diff
@@ -195,16 +203,43 @@ public void signBy(DarcId darcId, List<Signer> signers) throws CothorityCryptoEx
         }
     }
 
+    /**
+     * This function derives an instance ID from the instruction with the given string. The derived instance ID if
```
Comment (kc1212) [inline external/java/src/main/java/ch/epfl/dedis/lib/omniledger/Instruction.java]: `s/if/is`, or just say "is the key of this transaction"

File: omniledger/service/service_test.go [hunk 5]
```diff
@@ -80,15 +80,18 @@ func TestService_AddTransaction_WithFailure(t *testing.T) {
 }
 
 func TestService_AddTransaction_WithFailure_OnFollower(t *testing.T) {
+	// for {
+	// 	log.Print("Starting test")
 	testAddTransaction(t, 1, true)
+	// }
```
Comment (kc1212) [inline omniledger/service/service_test.go]: remove these?

**Pull request merged**